### PR TITLE
chore(reth-consensus): `no_std` support

### PR DESCRIPTION
**Pull request by martinezjorge**
Towards #9478 

This PR adds `no_std` support for `reth-consensus.`

### CODE REVIEW

**Comment by martinezjorge** [comment]
@mattsse could I please get some feedback on this?

**Review by shekhirin — CHANGES_REQUESTED** [review]
This is the right direction, couple suggestions:
1. You can use `derive_more` to replicate the functionality of `thiserror` generation of `Display` implementations: https://jeltef.github.io/derive_more/derive_more/display.html#example-usage
2. There should be no need to change the `Display` implementation of `HeaderConsensusError`, so we can keep using the `Debug` for `SealedHeader`. The `?` here means "use `Debug` to format" https://doc.rust-lang.org/std/fmt/index.html#formatting-traits

**Comment by martinezjorge** [comment]
Thank you very much! I'll keep working on this approach.

**Comment by martinezjorge** [comment]
@shekhirin Thanks again for your feedback and suggestions, your approach was much simpler than what I was trying to do and helped me get unstuck really quickly. Let me know if there are anymore issues or changes needed.

File: crates/consensus/consensus/Cargo.toml
Comment (mattsse) [inline crates/consensus/consensus/Cargo.toml:19]:
```suggestion
thiserror-no-std = { workspace = true, default-features = false, optional = true }
```

Comment (mattsse) [inline crates/consensus/consensus/Cargo.toml:19]: do we even still need this @shekhirin

**Review by mattsse — COMMENTED** [review]
sorry @martinezjorge this review slipped through

**Comment by martinezjorge** [comment]
> sorry @martinezjorge this review slipped through

All good man!

File: crates/consensus/consensus/Cargo.toml
Comment (shekhirin) [reply crates/consensus/consensus/Cargo.toml]: i don't think so, we manually implement the error now

File: Cargo.toml
Comment (shekhirin) [inline Cargo.toml:282]: @mattsse is that okay here?

File: crates/consensus/consensus/src/lib.rs
Comment (shekhirin) [inline crates/consensus/consensus/src/lib.rs:407]: and in other places as well

```suggestion
#[display(fmt = "Consensus error: {_0}, Invalid header: {_1:?}")]
```

**Review by shekhirin — CHANGES_REQUESTED** [review]
We have more usages of `thiserror` in the `reth-consensus` crate, let's replace them with `derive_more::Display` in this PR too? They have a bunch of `#[from]` that will require manual `From` trait implementations, but it should be easy enough.

File: Cargo.toml
Comment (mattsse) [reply Cargo.toml:282]: this should be fine

**Comment by mattsse** [comment]
> We have more usages of `thiserror` in the `reth-consensus` crate, let's replace them with `derive_more::Display` in this PR too? They have a bunch of `#[from]` that will require manual `From` trait implementations, but it should be easy enough.

one thing to keep in mind for the manual error impl is that we need to override source fn where applicable (where we wrap other errors in variants)

File: crates/consensus/consensus/src/lib.rs
Comment (martinezjorge) [reply crates/consensus/consensus/src/lib.rs:407]: Will do!

File: crates/consensus/consensus/Cargo.toml
Comment (martinezjorge) [inline crates/consensus/consensus/Cargo.toml:23]: Does it make sense to keep std feature if its array is empty?

File: crates/consensus/consensus/src/lib.rs
Comment (martinezjorge) [inline crates/consensus/consensus/src/lib.rs:21]: Are these feature gates still necessary since the std feature array is empty?

File: crates/consensus/consensus/Cargo.toml
Comment (martinezjorge) [reply crates/consensus/consensus/Cargo.toml]: Removed

File: crates/consensus/consensus/src/lib.rs
Comment (martinezjorge) [reply crates/consensus/consensus/src/lib.rs]: Completed

**Comment by martinezjorge** [comment]
> > We have more usages of `thiserror` in the `reth-consensus` crate, let's replace them with `derive_more::Display` in this PR too? They have a bunch of `#[from]` that will require manual `From` trait implementations, but it should be easy enough.
> 
> one thing to keep in mind for the manual error impl is that we need to override source fn where applicable (where we wrap other errors in variants)

Are you saying that whenever we wrap other errors in error enum variants we need to implement from? Or are you making a different point? Trying to make sure I understand you.

File: crates/consensus/consensus/Cargo.toml
Comment (shekhirin) [reply crates/consensus/consensus/Cargo.toml:23]: yes, because other crates can use it to enable `std` behavior in this crate

File: crates/consensus/consensus/src/lib.rs
Comment (shekhirin) [reply crates/consensus/consensus/src/lib.rs:21]: https://github.com/paradigmxyz/reth/pull/9663#discussion_r1703946604

**Comment by shekhirin** [comment]
@martinezjorge you can run `cargo expand --manifest-path crates/errors/Cargo.toml error` to see what does `thiserror` macros expand to:
```rust
mod error {
    ...
    #[allow(unused_qualifications)]
    impl std::error::Error for RethError {
        fn source(&self) -> ::core::option::Option<&(dyn std::error::Error + 'static)> {
            use thiserror::__private::AsDynError as _;
            #[allow(deprecated)]
            match self {
                RethError::Execution { 0: transparent } => {
                    std::error::Error::source(transparent.as_dyn_error())
                }
                RethError::Consensus { 0: transparent } => {
                    std::error::Error::source(transparent.as_dyn_error())
                }
                RethError::Database { 0: transparent } => {
                    std::error::Error::source(transparent.as_dyn_error())
                }
                RethError::Provider { 0: transparent } => {
                    std::error::Error::source(transparent.as_dyn_error())
                }
                RethError::Canonical { 0: transparent } => {
                    std::error::Error::source(transparent.as_dyn_error())
                }
                RethError::Other { 0: transparent } => {
                    std::error::Error::source(transparent.as_dyn_error())
                }
            }
        }
    }
    ...
}
```

File: crates/consensus/consensus/Cargo.toml
Comment (martinezjorge) [reply crates/consensus/consensus/Cargo.toml:23]: gotcha

File: crates/consensus/consensus/src/lib.rs
Comment (shekhirin) [inline crates/consensus/consensus/src/lib.rs:130]: please move it below the enum and use `Option` directly without the full path

File: crates/consensus/consensus/src/lib.rs
Comment (shekhirin) [inline crates/consensus/consensus/src/lib.rs:416]: please move it below the struct declaration

**Review by shekhirin — CHANGES_REQUESTED** [review]
almost there

File: crates/consensus/consensus/src/lib.rs
Comment (martinezjorge) [reply crates/consensus/consensus/src/lib.rs]: of course, done

File: crates/consensus/consensus/src/lib.rs
Comment (martinezjorge) [reply crates/consensus/consensus/src/lib.rs]: done

**Comment by martinezjorge** [comment]
> @martinezjorge you can run `cargo expand --manifest-path crates/errors/Cargo.toml error` to see what does `thiserror` macros expand to:

this was really helpful thanks

File: crates/consensus/consensus/src/lib.rs
Comment (martinezjorge) [reply crates/consensus/consensus/src/lib.rs]: whoops missed one

**Comment by martinezjorge** [comment]
@shekhirin I think this is good to go if you can please take a look when you get a chance

**Review by mattsse — APPROVED** [review]
lgtm, but this is blocked by https://github.com/paradigmxyz/reth/pull/10048

File: crates/evm/execution-errors/Cargo.toml
Comment (martinezjorge) [inline crates/evm/execution-errors/Cargo.toml:29]: What affect does this have overall?

Comment (shekhirin) [reply crates/evm/execution-errors/Cargo.toml:29]: enables the `std` feature of `reth-consensus`, otherwise this errors
```console
error[E0277]: the trait bound `ConsensusError: StdError` is not satisfied
   --> crates/evm/execution-errors/src/lib.rs:209:66
    |
209 |             Self::Consensus(source) => std::error::Error::source(source),
    |                                        ------------------------- ^^^^^^ the trait `StdError` is not implemented for `ConsensusError`
    |                                        |
    |                                        required by a bound introduced by this call

For more information about this error, try `rustc --explain E0277`.
error: could not compile `reth-execution-errors` (lib) due to 1 previous error
```

**Comment by shekhirin** [comment]
There are still some errors left in `reth-network-p2p` due to changes of `reth-consensus` error types https://github.com/paradigmxyz/reth/actions/runs/10359250687/job/28675253949?pr=9663

**Review by shekhirin — APPROVED** [review]

File: crates/evm/execution-errors/Cargo.toml
Comment (shekhirin) [reply crates/evm/execution-errors/Cargo.toml:29]: I ran `cargo c -p reth-execution-errors --no-default-features`. Changes in `reth-consensus` affected other crates, such as `reth-execution-errors`

Comment (martinezjorge) [reply crates/evm/execution-errors/Cargo.toml:29]: ohhhhhhhhh ok I see

**Pull request merged**
